crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -644,235 +722,248 @@ impl BlankLinesChecker {
-                    )));
-                } else {
-                    diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
-                        stylist.line_ending().repeat(BLANK_LINES_TOP_LEVEL as usize),
-                        locator.line_start(self.last_non_comment_line_end),
-                    )));
-                }
+                && line.kind.is_class_function_or_decorator()
+        {
+            // E302
+            let mut diagnostic = Diagnostic::new(
+                BlankLinesTopLevel {
+                    actual_blank_lines: line.preceding_blank_lines.count(),
+                    expected_blank_lines: expected_blank_lines_before_definition,
+                },
+                line.first_token_range,
+            );
 
-                diagnostics.push(diagnostic);
+            if let Some(blank_lines_range) = line.blank_lines.range() {
+                diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                    self.stylist
+                        .line_ending()
+                        .repeat(expected_blank_lines_before_definition as usize),
+                    blank_lines_range,
+                )));
+            } else {
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    self.stylist
+                        .line_ending()
+                        .repeat(expected_blank_lines_before_definition as usize),
+                    self.locator.line_start(state.last_non_comment_line_end),
+                )));
             }
 
-            let expected_blank_lines = if line.indent_length > 0 {
-                BLANK_LINES_METHOD_LEVEL
-            } else {
-                BLANK_LINES_TOP_LEVEL
-            };
+            diagnostics.push(diagnostic);
+        }
 
-            if line.blank_lines > expected_blank_lines {
-                // E303
-                let mut diagnostic = Diagnostic::new(
-                    TooManyBlankLines {
-                        actual_blank_lines: line.blank_lines.count(),
-                    },
-                    line.first_token_range,
-                );
+        let max_lines_level = if line.indent_length == 0 {
+            BLANK_LINES_TOP_LEVEL
+        } else {
+            BLANK_LINES_NESTED_LEVEL
+        };
+
+        // If between `import` and `from..import` or the other way round,
+        // allow up to `lines_between_types` newlines for isort compatibility.
+        // We let `isort` remove extra blank lines when the imports belong
+        // to different sections.
+        let max_blank_lines = if matches!(
+            (line.kind, state.follows),
+            (LogicalLineKind::Import, Follows::FromImport)
+                | (LogicalLineKind::FromImport, Follows::Import)
+        ) {
+            if self.lines_between_types == 0 {
+                max_lines_level
+            } else {
+                u32::try_from(self.lines_between_types).unwrap_or(u32::MAX)
+            }
```

Comment:
Just to clarify, this means that blank lines is more lenient than isort if `lines_between_types == 0`, but uses the value from isort otherwise. Which means that `E303` can be "stricter" (only accept a single blank line) 
when `lines_between_types == 1` than when  `lines_between_types == 0` ?

I'm guessing that this is to avoid blank lines triggering on something formatted by isort if `lines_between_types > max_lines_level`. Wouldn't it be simpler to use whichever value is highest ?

```suggestion
            std::cmp::max(
                max_lines_level,
                u32::try_from(self.lines_between_types).unwrap_or(u32::MAX),
            )
```

Reply:
Yeah, I think that's a very reasonable simplification to make. The rule doesn't need to match `isort` exactly, it just need to accept isorted code